tensorflow/python/grappler/auto_mixed_precision_test.py

```diff
@@ -497,37 +511,30 @@ def testRecurrentLSTM(self):
       self._assert_output_fp16(node_map, 'while/Tanh_1')
       self.assertAllClose(output_val_ref, output_val, atol=1e-3, rtol=1e-3)
 
-  def testPropagationThroughIntertwinedLoop1(self):
-    self._run_intertwined_loop_test('C', 'C', 'bgW', 'C', 'g', 'b', 4, 3)
-
-  def testPropagationThroughIntertwinedLoop2(self):
-    # Note that this results in NextIteration and Merge being painted different
-    # colors, requiring NextIteration to be forced to match.
-    self._run_intertwined_loop_test('b', 'g', 'gW', 'C', 'c', 'C', 3, 2)
+  def test_propagation_through_simple_loop_1(self):
+    self._run_simple_loop_test('W', 'C', 'C')
 
-  def testPropagationThroughIntertwinedLoop3(self):
-    self._run_intertwined_loop_test('g', 'g', 'g', 'g', 'W', 'c', 3, 2)
+  def test_propagation_through_simple_loop_2(self):
+    self._run_simple_loop_test('C', 'C', 'W')
 
-  def testPropagationThroughIntertwinedLoop4(self):
-    self._run_intertwined_loop_test('W', 'g', 'g', 'g', 'g', 'g', 3, 2)
+  def test_propagation_through_simple_loop_3(self):
+    self._run_simple_loop_test('W', 'G', 'W')
 
-  def testPropagationThroughIntertwinedLoop5(self):
-    self._run_intertwined_loop_test('W', 'c', 'b', 'c', 'c', 'W', 4, 2)
+  def test_propagation_through_simple_loop_4(self):
+    self._run_simple_loop_test('W', 'gbg', 'W')
 
-  def testPropagationThroughIntertwinedLoop6(self):
-    self._run_intertwined_loop_test('b', 'g', 'g', 'g', 'g', 'W', 2, 1)
+  def test_propagation_through_simple_loop_5(self):
+    self._run_simple_loop_test('b', 'gWC', 'c')
 
-  def testPropagationThroughIntertwinedLoop7(self):
-    self._run_intertwined_loop_test('c', 'c', 'bWg', 'c', 'g', 'b', 2, 1)
+  def test_propagation_through_simple_loop_6(self):
+    self._run_simple_loop_test('b', 'CWCG', 'C')
```

Comment:
Is the gray node made fp16 because it is "between" two whites nodes, where both white nodes are the same due to the loop? That sounds sensible, I am just checking.

Reply:
That's correct; the black does not propagate into the loop because the clear nodes after the blacklisted node are not upstream of any blacklisted/greylisted nodes.  Then the traversal tags the grey node (and clear nodes) within the loop as both upstream and downstream of a whitelisted node (in this case the only one).